circleci: failed tests (12973): TestGossipStorage

The following test appears to have failed:

[#12973](https://circleci.com/gh/cockroachdb/cockroach/12973):

```
I0218 05:35:30.369262 204 gossip/server.go:252  @1: distantHops: 2 from 4
I0218 05:35:30.370445 204 gossip/server.go:197  node 1 received [127.0.0.1:46108] from node 3 (0 fresh)
I0218 05:35:30.370591 204 gossip/server.go:252  @1: distantHops: 2 from 4
I0218 05:35:30.371367 204 gossip/server.go:197  node 2 received [127.0.0.1:43752] from node 4 (1 fresh)
I0218 05:35:30.371811 204 gossip/server.go:252  @2: distantHops: 2 from 3
--- FAIL: TestGossipStorage (0.16s)
	storage_test.go:120: 1: expected addresses: [{tcp 127.0.0.1:46108} {tcp 127.0.0.1:46143}], got: [{tcp 127.0.0.1:34993} {tcp 127.0.0.1:46108} {tcp 127.0.0.1:46143}]
FAIL
FAIL	github.com/cockroachdb/cockroach/gossip	1.604s
=== RUN   TestParseResolverSpec
--- PASS: TestParseResolverSpec (0.00s)
=== RUN   TestGetAddress
--- PASS: TestGetAddress (0.00s)
PASS
ok  	github.com/cockroachdb/cockroach/gossip/resolver	1.031s
?   	github.com/cockroachdb/cockroach/gossip/simulation	[no test files]
=== RUN   TestKeySorting
--- PASS: TestKeySorting (0.00s)
=== RUN   TestMakeKey

```
Please assign, take a look and update the issue accordingly.

--- a/gossip/gossip.go
+++ b/gossip/gossip.go
@@ -267,7 +267,7 @@ func (g *Gossip) SetStorage(storage Storage) error {
 	}
 
 	// Merge the stored bootstrap info addresses with any we've become
-	// aware of through the --join bootstrap hosts we've connected to.
+	// aware of through gossip.
 	if len(g.bootstrapInfo.Addresses) > 0 {
 		existing := map[string]struct{}{}
 		makeKey := func(a util.UnresolvedAddr) string { return fmt.Sprintf("%s,%s", a.Network(), a.String()) }

--- a/gossip/simulation/network.go
+++ b/gossip/simulation/network.go
@@ -105,6 +105,8 @@ func (n *Network) CreateNode() (*Node, error) {
 // StartNode initializes a gossip instance for the simulation node and
 // starts it.
 func (n *Network) StartNode(node *Node) error {
+	node.Gossip.Start(node.Server, node.Addr)
+	node.Gossip.EnableSimulationCycler(true)
 	n.nodeIDAllocator++
 	node.Gossip.SetNodeID(n.nodeIDAllocator)
 	if err := node.Gossip.SetNodeDescriptor(&roachpb.NodeDescriptor{
@@ -117,8 +119,6 @@ func (n *Network) StartNode(node *Node) error {
 		encoding.EncodeUint64Ascending(nil, 0), time.Hour); err != nil {
 		return err
 	}
-	node.Gossip.Start(node.Server, node.Addr)
-	node.Gossip.EnableSimulationCycler(true)
 	return nil
 }
 

--- a/gossip/storage_test.go
+++ b/gossip/storage_test.go
@@ -17,6 +17,7 @@
 package gossip_test
 
 import (
+	"errors"
 	"reflect"
 	"sort"
 	"strings"
@@ -37,31 +38,37 @@ type testStorage struct {
 	info        gossip.BootstrapInfo
 }
 
-func (tp *testStorage) isRead() bool {
-	tp.Lock()
-	defer tp.Unlock()
-	return tp.read
+func (ts *testStorage) isRead() bool {
+	ts.Lock()
+	defer ts.Unlock()
+	return ts.read
 }
 
-func (tp *testStorage) isWrite() bool {
-	tp.Lock()
-	defer tp.Unlock()
-	return tp.write
+func (ts *testStorage) isWrite() bool {
+	ts.Lock()
+	defer ts.Unlock()
+	return ts.write
 }
 
-func (tp *testStorage) ReadBootstrapInfo(info *gossip.BootstrapInfo) error {
-	tp.Lock()
-	defer tp.Unlock()
-	tp.read = true
-	*info = *util.CloneProto(&tp.info).(*gossip.BootstrapInfo)
+func (ts *testStorage) Len() int {
+	ts.Lock()
+	defer ts.Unlock()
+	return len(ts.info.Addresses)
+}
+
+func (ts *testStorage) ReadBootstrapInfo(info *gossip.BootstrapInfo) error {
+	ts.Lock()
+	defer ts.Unlock()
+	ts.read = true
+	*info = *util.CloneProto(&ts.info).(*gossip.BootstrapInfo)
 	return nil
 }
 
-func (tp *testStorage) WriteBootstrapInfo(info *gossip.BootstrapInfo) error {
-	tp.Lock()
-	defer tp.Unlock()
-	tp.write = true
-	tp.info = *util.CloneProto(info).(*gossip.BootstrapInfo)
+func (ts *testStorage) WriteBootstrapInfo(info *gossip.BootstrapInfo) error {
+	ts.Lock()
+	defer ts.Unlock()
+	ts.write = true
+	ts.info = *util.CloneProto(info).(*gossip.BootstrapInfo)
 	return nil
 }
 
@@ -100,6 +107,16 @@ func TestGossipStorage(t *testing.T) {
 	// Wait for the gossip network to connect.
 	network.RunUntilFullyConnected()
 
+	// Wait long enough for storage to get the expected number of addresses.
+	util.SucceedsWithin(t, 500*time.Millisecond, func() error {
+		for _, p := range stores {
+			if p.Len() != 2 {
+				return errors.New("incorrect number of addresses")
+			}
+		}
+		return nil
+	})
+
 	for i, p := range stores {
 		if !p.isRead() {
 			t.Errorf("%d: expected read from storage", i)
@@ -108,17 +125,22 @@ func TestGossipStorage(t *testing.T) {
 			t.Errorf("%d: expected write from storage", i)
 		}
 
+		p.Lock()
 		gotAddresses := unresolvedAddrSlice(p.info.Addresses)
-
 		sort.Sort(gotAddresses)
-		expectedAddresses := append(unresolvedAddrSlice(nil), addresses...)
-		expectedAddresses = append(expectedAddresses[:i], expectedAddresses[i+1:]...)
+		var expectedAddresses unresolvedAddrSlice
+		for j, addr := range addresses {
+			if i != j { // skip node's own address
+				expectedAddresses = append(expectedAddresses, addr)
+			}
+		}
 		sort.Sort(expectedAddresses)
 
 		// Verify all gossip addresses are written to each persistent store.
 		if !reflect.DeepEqual(gotAddresses, expectedAddresses) {
 			t.Errorf("%d: expected addresses: %s, got: %s", i, expectedAddresses, gotAddresses)
 		}
+		p.Unlock()
 	}
 
 	// Create an unaffiliated gossip node with only itself as a resolver,
@@ -148,10 +170,11 @@ func TestGossipStorage(t *testing.T) {
 
 	// Give the new node storage with info established from a node
 	// in the established network.
-	tp := &testStorage{
-		info: stores[0].info,
+	var ts2 testStorage
+	if err := stores[0].ReadBootstrapInfo(&ts2.info); err != nil {
+		t.Fatal(err)
 	}
-	if err := node.Gossip.SetStorage(tp); err != nil {
+	if err := node.Gossip.SetStorage(&ts2); err != nil {
 		t.Fatal(err)
 	}
 

